p2p/src/main/kotlin/maru/p2p/P2PNetworkFactory.kt

```diff
@@ -0,0 +1,165 @@
+  ): P2PNetwork<Peer> {
+    val rpcMethod = MaruRpcMethod()
+    val gossipTopicHandlers = GossipTopicHandlers()
+//    gossipTopicHandlers.add("topic", TestTopicHandler()) // TODO: I don't think this is needed
+
+    val gossipParams = GossipParamsBuilder().heartbeatInterval(100.millis).build()
+    val gossipRouterBuilder =
+      GossipRouterBuilder().apply {
+        params = gossipParams
+        scoreParams = GossipScoreParams(GossipPeerScoreParams(), GossipTopicsScoreParams())
+        messageFactory = { getMessageFactory(it, gossipTopicHandlers) }
+      }
+    val gossipRouter = gossipRouterBuilder.build()
+    val pubsubApiImpl = PubsubApiImpl(gossipRouter)
+    val gossip = Gossip(gossipRouter, pubsubApiImpl)
+
+    val metricsSystem = NoOpMetricsSystem()
+    val publisherApi = gossip.createPublisher(privateKey, Random.nextLong())
+    val gossipNetwork = LibP2PGossipNetwork(metricsSystem, gossip, publisherApi, gossipTopicHandlers)
+
+    val peerId = PeerId.fromPubKey(privateKey.publicKey())
+    val libP2PNodeId = LibP2PNodeId(peerId)
+
+    val metricTrackingExecutorFactory = MetricTrackingExecutorFactory(metricsSystem)
+    val asyncRunner = AsyncRunnerFactory.createDefault(metricTrackingExecutorFactory).create("maru", 2)
+    val rpcHandler = RpcHandler(asyncRunner, rpcMethod)
+
+    val peerManager =
+      PeerManager(
+        metricsSystem,
+        ReputationManager.NOOP,
+        mutableListOf<PeerHandler>(MaruPeerHandler()),
+        mutableListOf(rpcHandler),
+        { _ -> 50.0 }, // TODO: I guess we need a scoring function here
+      )
+
+    val host = createHost(privateKey, listOf(gossip, peerManager), gossip, rpcHandler, ipv4Address)
+
+    val advertisedAddresses =
+      mutableListOf<Multiaddr>().apply {
+        add(ipv4Address)
+      }
+
+    return LibP2PNetwork(
+      privateKey,
+      libP2PNodeId,
+      host,
+      peerManager,
+      advertisedAddresses,
+      gossipNetwork,
+      mutableListOf(1),
+    )
+  }
+
+  private fun getMessageFactory(
+    msg: Rpc.Message,
+    gossipTopicHandlers: GossipTopicHandlers,
+  ): PreparedPubsubMessage {
+    val arrivalTimestamp = Optional.empty<UInt64>()
+    val topic = msg.getTopicIDs(0)
+    val payload = Bytes.wrap(msg.data.toByteArray())
+
+    val preparedMessage =
+      gossipTopicHandlers
+        .getHandlerForTopic(topic)
+        .map { handler -> handler.prepareMessage(payload, arrivalTimestamp) }
+        .orElse(MaruPreparedGossipMessage(payload, arrivalTimestamp))
+
+    return PreparedPubsubMessage(msg, preparedMessage)
+  }
+
+  private fun createHost(
+    privKey: PrivKey,
+    connectionHandlers: List<ConnectionHandler>,
+    gossip: Gossip,
+    rpcHandler: RpcHandler<MaruOutgoingRpcRequestHandler, Bytes, MaruRpcResponseHandler>,
+    ipv4Address: Multiaddr,
+  ): Host {
+    val listenAddrs =
+      mutableListOf<String>().apply {
+        add(ipv4Address.toString())
+      }
+
+    return hostJ(Standard) { b: BuilderJ ->
```

Comment:
Do we have to use hostJ? libp2p has a Kotlin native builder

Reply:
using the kotlin builder now